oci/layer/utils.go

```diff
@@ -45,24 +48,49 @@ type MapOptions struct {
 // container mappings. Returns an error if it's not possible to map the given
 // UID.
 func mapHeader(hdr *tar.Header, mapOptions MapOptions) error {
-	// If we're in rootless mode, we assume all of the files are owned by
-	// (0, 0) in the container -- since we cannot map any other users.
-	if mapOptions.Rootless {
-		hdr.Uid, _ = idtools.ToHost(0, mapOptions.UIDMappings)
-		hdr.Gid, _ = idtools.ToHost(0, mapOptions.GIDMappings)
-	}
+	var newUid, newGid int
 
-	newUID, err := idtools.ToContainer(hdr.Uid, mapOptions.UIDMappings)
-	if err != nil {
-		return errors.Wrap(err, "map uid to container")
+	// It only makes sense to do un-mapping if we're not rootless. If we're
+	// rootless then all of the files will be owned by us anyway.
+	if !mapOptions.Rootless {
+		var err error
+		newUid, err = idtools.ToContainer(hdr.Uid, mapOptions.UIDMappings)
+		if err != nil {
+			return errors.Wrap(err, "map uid to container")
+		}
+		newGid, err = idtools.ToContainer(hdr.Gid, mapOptions.GIDMappings)
+		if err != nil {
+			return errors.Wrap(err, "map gid to container")
+		}
 	}
-	newGID, err := idtools.ToContainer(hdr.Gid, mapOptions.GIDMappings)
-	if err != nil {
-		return errors.Wrap(err, "map gid to container")
+
+	// We have special handling for the "user.rootlesscontainers" xattr. If
+	// we're rootless then we override the owner of the file we're currently
+	// parsing (and then remove the xattr). If we're not rootless then the user
+	// is doing something strange, so we log a warning but just ignore the
+	// xattr otherwise.
+	if value, ok := hdr.Xattrs[rootlesscontainers.Keyname]; !ok {
+		// noop
+	} else if !mapOptions.Rootless {
+		log.Warnf("suspicious filesystem: saw special rootless xattr %s in non-rootless invocation", rootlesscontainers.Keyname)
+	} else {
+		var payload rootlesscontainers.Resource
+		if err := proto.Unmarshal([]byte(value), &payload); err != nil {
+			return errors.Wrap(err, "unmarshal rootlesscontainers payload")
+		}
+
+		// If the payload isn't uint32(-1) we apply it. The xattr includes the
+		// *in-container* owner so we don't want to map it.
+		if uid := payload.GetUid(); uid != rootlesscontainers.NoopId {
+			newUid = int(uid)
+		}
+		if gid := payload.GetGid(); gid != rootlesscontainers.NoopId {
+			newGid = int(gid)
+		}
```

Comment:
I think `hdr.Xattrs[rootlesscontainers.Keyname]` should be removed here

Reply:
You're right -- the code used to do that but I must've lost it in a refactor. The tests I'm working on would've picked this up (for what it's worth).